File: 0.1.0/layer/dense.py
import numpy as np
from tools.vectorize import Vectorize

class Dense:

    def __init__(self, input_size = 0, layer_size = 0, randomization = True, activation_function = None):
        self.input_size = input_size
        self.layer_size = layer_size
        self.activation_function = activation_function
        if randomization:
            self.weights = 0.1*np.random.randn(input_size, layer_size)
            self.biases = 0.1*np.random.randn(layer_size)
        else:
            self.weights = np.zeros((input_size, layer_size))
            self.biases = np.zeros(layer_size)

    def run(self, input):
        self.value = np.matmul(input, self.weights) + self.biases
        self.output = Vectorize(self.activation_function.function, matrix_a=self.value)
        self.propagation_units = np.zeros((self.input_size, self.layer_size, np.shape(input)[0]))

File: 0.2.0/layer/dense.jl
module dense
    mutable struct Dense
        activator::Any

        input_size::Int64
        layer_size::Int64
        activation_function::Module

        weights::Array{Float32}
        biases::Array{Float32}

        value::Array{Float32}
        output::Array{Float32}
        propagation_units::Array{Float32}

        function Dense(;input_size::Int64, layer_size::Int64, randomization::Bool, activation_function::Module, reload::Bool=false)
            if reload
                return new(activate_Dense)
            end

            if randomization
                weights = 0.1*rand(Float32, (layer_size, input_size))
                biases = 0.1*rand(Float32, layer_size)
            else
                weights = zeros(Float32, (layer_size, input_size))
                biases = zeros(Float32, layer_size)
            end
            new(activate_Dense, input_size, layer_size, activation_function, weights, biases)
        end
    end

    function activate_Dense(layer::Dense, input::Array{Float32})
        layer.value = layer.weights*input .+ layer.biases
        layer.output = layer.activation_function.func(layer.value)
        layer.propagation_units = zeros((layer.layer_size, layer.input_size, size(input, 2)))
    end
end
